### PR TITLE
Add virtio-iommu support

### LINKED ISSUE #314: Add virtio-iommu support
The goal of supporting virtio-iommu is to expose a virtual IOMMU to the guest and attach both virtio and VFIO devices to it.
The point is to increase security regarding those devices, but also to be able to pass a real device (VFIO) through multiple layers of virtualization.

__TODO:__
- [x] Implement `virtio-iommu` device
- [x] Implement for `virtio-pci`
- [x] Implement for `virtio-mmio`
  - Actually impossible to do since `virtio-iommu` has a dependency on ACPI/DT, and the current configuration for cloud-hypervisor is that `virtio-mmio` does not support ACPI.
- [ ] Implement for `VFIO`
- [x] Expose an IORT table with the correct mappings
- [x] Translate IOVA addresses into GPA for virtio devices behind the vIOMMU 
- [ ] Implement `reset()` for all virtio devices
  - `virtio-blk` and `virtio-net` are ready
- [ ] Add `iommu=on|off` option for all virtio devices
  - `virtio-blk` and `virtio-net` are ready
- [x] Cleanup commits
- [x] Add documentation for virtual IOMMU
- [ ] Update kernel branch to include x86 patches for virtio-iommu
- [ ] More integration tests

### PR DESCRIPTION
**Pull request by sboeuf**
This pull request intends to bring support for virtual IOMMU relying on the brand new `virtio-iommu` device specification.

Fixes #314

### CODE REVIEW

**Comment by sboeuf** [comment]
@rbradford @sameo 
Just wanted to give some update here. I've been able, based on the patches from @eauger and @jpbrucker, to boot a VM with a `virtio-iommu-pci` device along with a paravirtualized IOMMU and a PCI root complex IORT nodes. Altogether, this exposes to the guest some specific devices that are described in the PCI root complex node, as attached to the IOMMU. That's how we can see some devices being part of iommu groups under `/sys/kernel/iommu_groups/`.
Because of this, I could use VFIO to bind the device that I had identified as being attached to the IOMMU, and then start a nested VM from this first VM.
And here are the two use cases, either we're passing a virtio device, or we're passing a real device that has been already passed through.
This implies a lot of inter dependency between virtio devices, VFIO devices and the virtio-iommu device.
- In case of virtio, we need to implement the `VIRTIO_F_IOMMU_PLATFORM` feature, which implies to translate addresses coming from the guest driver before to try reading from the guest memory. Indeed, if we virtually attach a virtio device to a vIOMMU, we cannot assume anymore that the guest address is a physical address. It will be an IOVA, which is why we need to handle it from any virtio device that would support this feature. That means we need to store the mapping provided by the `virtio-iommu` device, and use it to perform the translation.
- In case of VFIO, we want to call `VFIO_DMA_MAP` ioctl on the existing VFIO container, which will update the previous mapping in the physical IOMMU. This case is different as every time we receive a new MAP request through virtio-iommu, we need to trigger a VFIO ioctl call.

There's one last thing that is going to be challenging to tackle, we need to know the BDF of each device we're placing on the PCI bus, as we need this to provide it to the PCI root complex node, so that it can identify which devices are attached to the IOMMU. I think we need to build a list of the devices that will end up attached to the IOMMU, and once all devices have been initialized, we can generate the proper IORT table with the associated IORT node and mappings.

I wanted to throw my thoughts here as we need to have some design discussions on how to handle this :)

**Comment by rbradford** [comment]
> And here are the two use cases, either we're passing a virtio device, or we're passing a real device that has been already passed through.

Are you talking about nested VFIO here?  Is that a very frequent use case?

**Comment by sameo** [comment]
On Fri, Sep 27, 2019 at 06:12:45PM -0700, Sebastien Boeuf wrote:
> This implies a lot of inter dependency between virtio devices, VFIO devices and the virtio-iommu device.
> - In case of virtio, we need to implement the `VIRTIO_F_IOMMU_PLATFORM` feature, which implies to translate addresses coming from the guest driver before to try reading from the guest memory. Indeed, if we virtually attach a virtio device to a vIOMMU, we cannot assume anymore that the guest address is a physical address. It will be an IOVA, which is why we need to handle it from any virtio device that would support this feature. That means we need to store the mapping provided by the `virtio-iommu` device, and use it to perform the translation.

I guess this is the price we have to pay for pretending to be a physical
device. Who exposes the `VIRTIO_F_IOMMU_PLATFORM` feature? I hope we can
decide if we need to do the translation at runtime or not. In case of
non directly assigned virtio devices, supporting virtio-iommu should
have zero impact.

> - In case of VFIO, we want to call `VFIO_DMA_MAP` ioctl on the existing VFIO container, which will update the previous mapping in the physical IOMMU. This case is different as every time we receive a new MAP request through virtio-iommu, we need to trigger a VFIO ioctl call.
>

I suspect this is not something that happens *that* often for actual
phyical devices, when they are managed by a sane guest driver.

> There's one last thing that is going to be challenging to tackle, we need to know the BDF of each device we're placing on the PCI bus, as we need this to provide it to the PCI root complex node, so that it can identify which devices are attached to the IOMMU. I think we need to build a list of the devices that will end up attached to the IOMMU, and once all devices have been initialized, we can generate the proper IORT table with the associated IORT node and mappings.
>

Hmm...Right now we have that knowledge, but I guess that would prevent
us from doing vfio+hotplug, i.e. directly assigning a PCI device while
the VM is running?

**Comment by sboeuf** [comment]
@rbradford 
>Are you talking about nested VFIO here? Is that a very frequent use case?

Well VFIO nested is only one subsequent application out of the more general use case. If we expose to the guest a virtio device that we attach to this vIOMMU, then we need to handle the IOVA->PA translation. And if we expose a real device that we passed through, and that we attach this device to the vIOMMU, then the driver (let's say e1000 in case we're dealing with a NIC) will start doing some IOVA allocations, which need to be known by our VMM. So those cases don't even involve nesting anything, it's just a case where the guest drivers behave differently because they can see the devices are attached to an IOMMU.

**Comment by sboeuf** [comment]
@sameo 
>I guess this is the price we have to pay for pretending to be a physical
device. Who exposes the `VIRTIO_F_IOMMU_PLATFORM` feature? I hope we can
decide if we need to do the translation at runtime or not. In case of
non directly assigned virtio devices, supporting virtio-iommu should
have zero impact.

Yes of course we can decide to expose the feature or not. But the point is to be able to expose the feature for any virtio device that we would attach to this vIOMMU.

>I suspect this is not something that happens *that* often for actual
phyical devices, when they are managed by a sane guest driver.

In case of VFIO, there are two stages. When the `e1000` driver is being loaded from the guest, there might be a bunch of random allocations. But once you unbind the device from the e1000 driver and bind it to VFIO, once the nested VMM instance uses VFIO, the entire guest L2 memory is pinned, which is why we get a bunch of 4k pages allocations for the entire range, but this happens only once.

>Hmm...Right now we have that knowledge, but I guess that would prevent
us from doing vfio+hotplug, i.e. directly assigning a PCI device while
the VM is running?

Oh we do have that knowledge?! That's great news then. Why do you think this might interfere with hotplug? In the case of vIOMMU, we don't need a specific device slot, we only need to anticipate where the device is going to end up from a PCI hierarchy perspective, that's it. If we already have a way to know this, then we should be fine with this issue.

**Comment by sboeuf** [comment]
@rbradford @sameo This PR is now ready for a round of review. I cleaned up the commits and added a very simple integration test. This has support for both `virtio-blk` and `virtio-net`.

The next steps are:
- Having a reliable kernel branch with the patches (waiting for virtio-iommu maintainers to get some updates)
- Based on this new kernel, add some tests where using `virtio-blk` and `virtio-net` with the option `iommu=on`
- Update the VFIO integration test, replacing QEMU with cloud-hypervisor 
- Implement iommu support for all remaining virtio devices
- Last but not the least, add the virtual iommu support to VFIO devices

Let me know how you want to proceed with those steps. Should we merge something first and come up with follow up PRs for all those bullet points?

**Comment by rbradford** [comment]
Please make sure the README is updated with the command line changes and the purpose of the feature and any impact turning it on might have (performance) and the reason why you would want to turn it.

File: vm-virtio/src/iommu.rs [hunk 1]
```diff
@@ -0,0 +1,820 @@
...
+#[allow(unused)]
+const VIRTIO_IOMMU_F_MAP_UNMAP: u32 = 2;
+#[allow(unused)]
+const VIRTIO_IOMMU_F_BYPASS: u32 = 3;
+#[allow(unused)]
+const VIRTIO_IOMMU_F_PROBE: u32 = 4;
+
+// Default to 4KiB page size.
+#[allow(unused)]
+const VIRTIO_IOMMU_PAGE_SIZE_MASK: u64 = 1 << 12;
```
Comment (rbradford) [inline vm-virtio/src/iommu.rs]: Personally, I prefer `4 << 10`

File: vmm/src/device_manager.rs [hunk 2]
```diff
@@ -383,8 +392,25 @@ impl DeviceManager {
...
+            let (iommu_device, mapping) =
+                vm_virtio::Iommu::new().map_err(DeviceManagerError::CreateVirtioIommu)?;
+            virtio_devices.push((
+                Box::new(iommu_device) as Box<dyn vm_virtio::VirtioDevice>,
+                false,
+            ));
+            Some(mapping)
+        } else {
+            None
+        };
```
Comment (rbradford) [inline vmm/src/device_manager.rs]: Could you move this down into the `pci_support` block below? Then you could get away without naming the variable with the _ prefix.

File: vmm/src/device_manager.rs [hunk 3]
```diff
@@ -563,22 +606,30 @@ impl DeviceManager {
...
                     ImageType::Raw => {
                         let raw_img = vm_virtio::RawFile::new(raw_img);
-                        let dev =
-                            vm_virtio::Block::new(raw_img, disk_cfg.path.clone(), false, false)
-                                .map_err(DeviceManagerError::CreateVirtioBlock)?;
+                        let dev = vm_virtio::Block::new(
+                            raw_img,
+                            disk_cfg.path.clone(),
+                            false,
+                            disk_cfg.iommu,
```
Comment (rbradford) [inline vmm/src/device_manager.rs]: What happens if a user turns it on for a disk and forgets to turn it on globally?

**Comment by rbradford** [comment]
Overall lgtm. I see no harm in merging a WIP, this PR does clash with #273

File: vmm/src/device_manager.rs (hunk 3 above)
Comment (sboeuf) [reply vmm/src/device_manager.rs]: The virtio feature flag is enabled in this case, but because the virtio-iommu is not created, it has strictly no impact on the way the device works, with no impact on performance.

File: vmm/src/device_manager.rs (hunk 2 above)
Comment (sboeuf) [reply vmm/src/device_manager.rs]: Yes good point!

**Comment by sboeuf** [comment]
@sameo @rbradford All comments addressed, and the PR is already rebased on #273, so it's ready to be merged whenever #273 gets merged.

**Comment by sameo** [comment]
Rebased on top of master.

File: docs/iommu.md [hunk 4]
```diff
@@ -0,0 +1,87 @@
...
+attached to the virtual IOMMU.
+
+```bash
+./cloud-hypervisor \
+    --cpus 1 \
+    --memory size=512M \
+    --disk path=clear-kvm.img,iommu=on \
+    --kernel custom-bzImage \
+    --cmdline "console=ttyS0 root=/dev/vda3" \
+    --iommu
```
Comment (sameo) [inline docs/iommu.md]: Could we have a simpler user experience by implicitly enabling the virtio IOMMU if any of the devices (parvirt'ed or directly assigned) asks for going through the IOMMU.
The `--iommu` case would still make sense for when we'll want to do device hotplug and have it go through our IOMMU, although I guess we might also dynamically add it.
Bottom line is: can we make the `--iommu` cli switch optional?

File: docs/iommu.md [hunk 5]
```diff
@@ -0,0 +1,87 @@
...
+case of nested virtualization as it can protect the first layer guest from
+malicious guests running in L2. 
+
+As suggested by the above statement, having a virtual IOMMU also allows for
+nested device passthrough. Let's say the system has a physical IOMMU, it is
+now possible to expose a virtual IOMMU to the guest, which will be responsible
+for updating the physical DMA remapping table. This will happen through the
+VFIO framework, and will eventually allow for passing a physical device through
+multiple layers of virtualization.
+
```
Comment (sameo) [inline docs/iommu.md]: Another question that this documentation triggers: if I do a device passthrough with `--iommu`, I assume the guest driver will switch to using IOVA and implictly expect our virtual IOMMU to do the IOVA->PA translation. I see that this PR passes the VFIO CI tests, though: Any idea how that works?

File: docs/iommu.md [hunk 6]
```diff
@@ -0,0 +1,87 @@
...
+interesting to leverage for some use cases.
+
+The first reason why one might want to expose a virtual IOMMU to the
+guest is to increase the security regarding the memory accesses performed
+by the virtual devices, on behalf of the guest drivers. With a virtual
+IOMMU, the VMM can use interposition to make choices about letting a device
+access the guest memory, the same way a physical IOMMU would prevent a
+physical device from accessing the physical memory. The security gain is only
+visible from a guest perspective, which does not apply if the guest we're
+taking as example is corrupted. The benefits about security can be seen in
```
[the remark this answers is not in the document]
Comment (sboeuf) [reply docs/iommu.md]: Yes `malicious` is the right word.

File: docs/iommu.md (hunk 5 above)
Comment (sboeuf) [reply docs/iommu.md]: Because we have to choose which devices are attached to the vIOMMU. By default, a device is not attached to it, therefore the driver does not use IOVAs.

File: docs/iommu.md (hunk 4 above)
Comment (sboeuf) [reply docs/iommu.md]: Yes you're right, I changed the code to enable virtio-iommu creation if at least one of the devices has `iommu=on` flag.

File: docs/iommu.md [hunk 7]
```diff
@@ -0,0 +1,87 @@
...
+case of nested virtualization as it can protect the first layer guest from
+malicious guests running in L2.
```
[the remark this answers is not in the document]
Comment (sboeuf) [reply docs/iommu.md]: Yes let me rephrase all this properly.

**Comment by sboeuf** [comment]
@sameo All comments addressed, PTAL
